core/optaplanner-constraint-streams/src/main/java/org/optaplanner/constraint/streams/bavet/common/NodeBuildHelper.java

```diff
@@ -46,9 +46,14 @@ public NodeBuildHelper(Set<? extends ConstraintStream> activeStreamSet,
             Map<Constraint, Score_> constraintWeightMap,
             AbstractScoreInliner<Score_> scoreInliner) {
         this.activeStreamSet = activeStreamSet;
-        insertMap = new HashMap<>(Math.max(16, activeStreamSet.size()));
-        retractMap = new HashMap<>(Math.max(16, activeStreamSet.size()));
-        storeIndexMap = new HashMap<>(Math.max(16, activeStreamSet.size() / 2));
+        /*
+         * Nodes are only being built after the streams have already been node-shared.
+         * If two streams are node-shared, they have by this point already been reduced to a single stream instance.
+         * Therefore there is no need to use stream equality.
```

Comment:
That is true.
But if we use HashMap, does it work?

Reply:
It should. But see my explanation below one of your earlier comments. The same rationale applies here.